virt_lightning/api.py

```diff
@@ -0,0 +1,442 @@
+#!/usr/bin/env python3
+
+import collections
+from concurrent.futures import ThreadPoolExecutor
+import logging
+
+import asyncio
+import libvirt
+import re
+import pathlib
+
+import urllib.request
+import sys
+import distutils.util
+
+from virt_lightning.symbols import get_symbols
+
+
+import virt_lightning.virt_lightning as vl
+
+logger = logging.getLogger("virt_lightning")
+
+
+def libvirt_callback(userdata, err):
+    pass
+
+
+libvirt.registerErrorHandler(f=libvirt_callback, ctx=None)
+symbols = get_symbols()
+
+MB = 1024 * 1000
+
+
+def _register_aio_virt_impl(loop):
+    # Ensure we may call shell.up() multiple times
+    # from the same asyncio program.
+    loop = loop or asyncio.get_event_loop()
+    if loop not in _register_aio_virt_impl.aio_virt_bindinds:
+        try:
+            import libvirtaio
+
+            libvirtaio.virEventRegisterAsyncIOImpl(loop=loop)
+        except ImportError:
+            libvirt.virEventRegisterDefaultImpl()
+        _register_aio_virt_impl.aio_virt_bindinds[loop] = True
+
+
+_register_aio_virt_impl.__dict__["aio_virt_bindinds"] = {}
+
+
+def _start_domain(hv, host, context, configuration):
+    if host["distro"] not in hv.distro_available():
+        logger.error("distro not available: %s", host["distro"])
+        logger.info(
+            "Please select on of the following distro: %s", hv.distro_available()
+        )
+        exit()
```

Comment:
I think this `exit()` should replaced to `raise Exception`.

Reply:
Yes, good point.